File: src/main/java/org/team1/nbe1_3_team01/global/config/SecurityConfig.java
package org.team1.nbe1_3_team01.global.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.ProviderManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.CsrfConfigurer;
import org.springframework.security.config.annotation.web.configurers.FormLoginConfigurer;
import org.springframework.security.config.annotation.web.configurers.HeadersConfigurer;
import org.springframework.security.config.annotation.web.configurers.HttpBasicConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.factory.PasswordEncoderFactories;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.logout.LogoutFilter;
import org.team1.nbe1_3_team01.domain.user.repository.UserRepository;
import org.team1.nbe1_3_team01.global.auth.login.filter.CustomUsernamePasswordAuthenticationFilter;
import org.team1.nbe1_3_team01.global.auth.jwt.filter.JwtAuthenticationProcessingFilter;
import org.team1.nbe1_3_team01.global.auth.login.handler.LoginFailureHandler;
import org.team1.nbe1_3_team01.global.auth.login.handler.LoginSuccessHandler;
import org.team1.nbe1_3_team01.global.auth.jwt.service.JwtService;
import org.team1.nbe1_3_team01.global.auth.login.service.LoginService;
import org.team1.nbe1_3_team01.global.auth.jwt.respository.RefreshTokenRepository;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {
    private final LoginService loginService;
    private final JwtService jwtService;
    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final ObjectMapper objectMapper;
    private final WebMvcConfig webMvcConfig;

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http
                .cors(cors -> cors.configurationSource(webMvcConfig.corsConfigurationSource()))
                .csrf(CsrfConfigurer<HttpSecurity>::disable)
                .formLogin(FormLoginConfigurer<HttpSecurity>::disable)
                .httpBasic(HttpBasicConfigurer<HttpSecurity>::disable)
                .headers(it -> it.frameOptions(HeadersConfigurer.FrameOptionsConfig::sameOrigin))
                .sessionManagement(it ->
                        it.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                )
                .authorizeHttpRequests(
                        authorize -> authorize
                                .requestMatchers( "/","/swagger-ui/**", "/v3/api-docs/**").permitAll()
                                .requestMatchers("/api/user/sign-up/**").permitAll()
                                .requestMatchers("/user/sign-up/**").permitAll()
                                .requestMatchers("/static/**").permitAll()
                                .requestMatchers("/templates/**").permitAll()
                                .requestMatchers("/api/user/admin/**").hasRole("ADMIN")
                                .requestMatchers("/api/email/admin/**").hasRole("ADMIN")
                                .requestMatchers("/api/course/admin/**").hasRole("ADMIN")
                                .requestMatchers("/api/teams/admin/**").hasRole("ADMIN")
                                .anyRequest().authenticated()

                );
         http.addFilterAfter(customUsernamePasswordAuthenticationFilter(), LogoutFilter.class);
         http.addFilterBefore(jwtAuthenticationProcessingFilter(), CustomUsernamePasswordAuthenticationFilter.class);
        return http.build();
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
    }

    @Bean
    public AuthenticationManager authenticationManager() {
        DaoAuthenticationProvider provider = new DaoAuthenticationProvider();
        provider.setPasswordEncoder(passwordEncoder());
        provider.setUserDetailsService(loginService);
        return new ProviderManager(provider);
    }

    @Bean
    public LoginSuccessHandler loginSuccessHandler() {
        return new LoginSuccessHandler(jwtService);
    }

    @Bean
    public LoginFailureHandler loginFailureHandler() {
        return new LoginFailureHandler();
    }

    @Bean
    JwtAuthenticationProcessingFilter jwtAuthenticationProcessingFilter() {
        return new JwtAuthenticationProcessingFilter(jwtService, userRepository, refreshTokenRepository);
    }

    @Bean
    CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter() {
        CustomUsernamePasswordAuthenticationFilter customUsernamePasswordAuthenticationFilter
                = new CustomUsernamePasswordAuthenticationFilter(objectMapper);
        customUsernamePasswordAuthenticationFilter.setAuthenticationManager(authenticationManager());
        customUsernamePasswordAuthenticationFilter.setAuthenticationSuccessHandler(loginSuccessHandler());
        customUsernamePasswordAuthenticationFilter.setAuthenticationFailureHandler(loginFailureHandler());
        return customUsernamePasswordAuthenticationFilter;
    }

}

File: src/main/java/org/team1/nbe1_3_team01/global/config/SecurityConfig.kt
package org.team1.nbe1_3_team01.global.config

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.ProviderManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.logout.LogoutFilter
import org.team1.nbe1_3_team01.domain.user.repository.UserRepository
import org.team1.nbe1_3_team01.global.auth.jwt.filter.JwtAuthenticationProcessingFilter
import org.team1.nbe1_3_team01.global.auth.jwt.respository.RefreshTokenRepository
import org.team1.nbe1_3_team01.global.auth.jwt.service.JwtService
import org.team1.nbe1_3_team01.global.auth.login.filter.CustomUsernamePasswordAuthenticationFilter
import org.team1.nbe1_3_team01.global.auth.login.handler.LoginFailureHandler
import org.team1.nbe1_3_team01.global.auth.login.handler.LoginSuccessHandler
import org.team1.nbe1_3_team01.global.auth.login.service.LoginService

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val loginService: LoginService,
    private val jwtService: JwtService,
    private val userRepository: UserRepository,
    private val refreshTokenRepository: RefreshTokenRepository,
    private val objectMapper: ObjectMapper,
    private val webMvcConfig: WebMvcConfig

) {
    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .cors { it.configurationSource(webMvcConfig.corsConfigurationSource()) }
            .csrf { it.disable() }
            .formLogin { it.disable() }
            .httpBasic { it.disable() }
            .headers { it.frameOptions { frameOptions -> frameOptions.sameOrigin() } }
            .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .authorizeHttpRequests {
                it
                    .requestMatchers("/", "/swagger-ui/**", "/v3/api-docs/**").permitAll()
                    .requestMatchers("/api/user/sign-up/**").permitAll()
                    .requestMatchers("/user/sign-up/**").permitAll()
                    .requestMatchers("/static/**").permitAll()
                    .requestMatchers("/templates/**").permitAll()
                    .requestMatchers("/api/user/admin/**").hasRole("ADMIN")
                    .requestMatchers("/api/email/admin/**").hasRole("ADMIN")
                    .requestMatchers("/api/course/admin/**").hasRole("ADMIN")
                    .requestMatchers("/api/teams/admin/**").hasRole("ADMIN")
                    .anyRequest().authenticated()
            }
        http.addFilterAfter(customUsernamePasswordAuthenticationFilter(), LogoutFilter::class.java)
        http.addFilterBefore(jwtAuthenticationProcessingFilter(), CustomUsernamePasswordAuthenticationFilter::class.java)
        return http.build()
    }

    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return PasswordEncoderFactories.createDelegatingPasswordEncoder()
    }

    @Bean
    fun authenticationManager(): AuthenticationManager {
        val provider = DaoAuthenticationProvider()
        provider.setPasswordEncoder(passwordEncoder())
        provider.setUserDetailsService(loginService)
        return ProviderManager(provider)
    }

    @Bean
    fun loginSuccessHandler(): LoginSuccessHandler {
        return LoginSuccessHandler(jwtService)
    }

    @Bean
    fun loginFailureHandler(): LoginFailureHandler {
        return LoginFailureHandler()
    }

    @Bean
    fun jwtAuthenticationProcessingFilter(): JwtAuthenticationProcessingFilter {
        return JwtAuthenticationProcessingFilter(jwtService, userRepository, refreshTokenRepository)
    }

    @Bean
    fun customUsernamePasswordAuthenticationFilter(): CustomUsernamePasswordAuthenticationFilter {
        val customUsernamePasswordAuthenticationFilter = CustomUsernamePasswordAuthenticationFilter(objectMapper)
        customUsernamePasswordAuthenticationFilter.setAuthenticationManager(authenticationManager())
        customUsernamePasswordAuthenticationFilter.setAuthenticationSuccessHandler(loginSuccessHandler())
        customUsernamePasswordAuthenticationFilter.setAuthenticationFailureHandler(loginFailureHandler())
        return customUsernamePasswordAuthenticationFilter
    }
}
